Source/FieldSolver/MagnetostaticSolver/MagnetostaticSolver.cpp

```diff
@@ -0,0 +1,462 @@
+
+#include <array>
+#include <memory>
+#include <string>
+
+using namespace amrex;
+
+void
+WarpX::ComputeCurrentDensityField()
+{
+    WARPX_PROFILE("WarpX::ComputeCurrentDensityField");
+    // Fields have been reset in Electrostatic solver for this time step, these fields
+    // are added into the E & B fields after electrostatic solve
+
+    WARPX_ALWAYS_ASSERT_WITH_MESSAGE(this->max_level == 0, "Magnetostatic solver not implemented with mesh refinement.");
+
+    AddCurrentDensityFieldLabFrame();
+}
+
+void
+WarpX::AddCurrentDensityFieldLabFrame()
+{
+    WARPX_PROFILE("WarpX::AddCurrentDensityFieldLabFrame");
+
+    // Store the boundary conditions for the field solver if they haven't been
+    // stored yet
+    if (!m_vector_poisson_boundary_handler.bcs_set) {
+        m_vector_poisson_boundary_handler.defineVectorPotentialBCs();
+    }
+
+#ifdef WARPX_DIM_RZ
+    WARPX_ALWAYS_ASSERT_WITH_MESSAGE(n_rz_azimuthal_modes == 1,
+                                     "Error: RZ magnetostatic only implemented for a single mode");
+#endif
+
+    // reset current_fp before depositing current density for this step
+    for (int lev = 0; lev <= max_level; lev++) {
+        for (int dim=0; dim < 3; dim++) {
+            current_fp[lev][dim]->setVal(0.);
+        }
+    }
+
+    // Deposit current density (source of Poisson solver)
+    for (int ispecies=0; ispecies<mypc->nSpecies(); ispecies++){
+        WarpXParticleContainer& species = mypc->GetParticleContainer(ispecies);
+        if (!species.do_not_deposit) {
+            species.DepositCurrent(current_fp, dt[0], 0.);
+        }
+    }
+
+    // This will interpolate to staggered grid, filter, and synchronize
+    SyncCurrent(current_fp, current_cp); // Apply filter, perform MPI exchange, interpolate across levels
+
+#ifdef WARPX_DIM_RZ
+    for (int lev = 0; lev <= max_level; lev++) {
+        ApplyInverseVolumeScalingToCurrentDensity(current_fp[lev][0].get(),
+                                                  current_fp[lev][1].get(),
+                                                  current_fp[lev][2].get(), lev);
+    }
+#endif
+
+    // set the boundary and current density potentials
+    setVectorPotentialBC(vector_potential_fp_nodal);
+
+    // Compute the vector potential A, by solving the Poisson equation
+    WARPX_ALWAYS_ASSERT_WITH_MESSAGE( !IsPythonCallBackInstalled("poissonsolver"),
+        "Python Level Poisson Solve not supported for Magnetostatic implementation.");
+
+    computeVectorPotential( current_fp, vector_potential_fp_nodal, self_fields_required_precision,
+                     self_fields_absolute_tolerance, self_fields_max_iters,
+                     self_fields_verbosity );
+}
+
+/* Compute the vector potential `A` by solving the Poisson equation with `J` as
+   a source.
+   This uses the amrex solver.
+
+    More specifically, this solves the equation
+    \f[
+        \vec{\nabla}^2 r \vec{A} - (\vec{\beta}\cdot\vec{\nabla})^2 r \vec{A} = - r \mu_0 \vec{J}
```

Comment:
`beta` is not used here.
```suggestion
        \vec{\nabla}^2 r \vec{A} = - r \mu_0 \vec{J}
```